Data/Primitive/ByteArray.hs

```diff
@@ -223,17 +231,20 @@ foldrByteArray f z arr = go 0
       | otherwise                    = z
     sz = sizeOf (undefined :: a)
 
-fromListN :: Prim a => Int -> [a] -> ByteArray
-fromListN n ys = runST $ do
+byteArrayFromList :: Prim a => [a] -> ByteArray
+byteArrayFromList xs = byteArrayFromListN (length xs) xs
```

Comment:
We have shrinking here, so we should probably benchmark an array doubling (or similar) option. Array doubling would let this be a "good consumer" for list fusion, as well as compacting the representation much more aggressively regardless.

Reply:
I tried to start on this, and I realized I don't actually know how to make something that runs in `ST` a good list consumer. I know it requires `foldr`, and I'll want to delay inlining. In this context, would the specialized type of `foldr` look like this:

    foldr :: Prim a => (a -> ST s (MutableByteArray s) -> ST s (MutableByteArray s)) -> ST s (MutableByteArray s) -> [a] -> ST s (MutableByteArray s)

And then perform the freeze and the `runST` after this. Is that right?

Reply:
Let's deal with this question in a separate issue.